### ISSUE TITLE
file browser regex box slow on large directories

### ISSUE THREAD

**Issue by gperciva** [Tarsnap/tarsnap-gui#44]
Backup -> file browser -> regex box

If I type (at my normal quick typing speed),

    /usr/bin/a

then there's a 21-second pause between `/usr/bin/` and `a`, while the app (presumably) reads filenames.  That pause has 100% CPU usage (in one core/thread); unknown disk usage (I could dig into this some more if you want).  After the first time I did that, it's much faster -- only a 2-second pause between `bin/` and `a`.  But it's still a noticeable pause.

Ubuntu 14.04, Intel(R) Core(TM) i5-2415M CPU @ 2.30GHz, 8 Gb RAM.

**Comment by gperciva** [comment]
I tried using `valgrind --callgrind` after compiling with `CONFIG += debug`.  According to that, 82% of the CPU time is spent in `QCollator::QCollator(Qlocal const&)` (including sub-functions), and 66% of the time is spent in `ucol_open_52` (including sub-functions), which belongs to `libicui18n.so.52.1`.  It's interesting to see that `QCollator::QCollator(Qlocal const&)` was called 125752 times.  `ucol_next_52` was called 27665660 times.  (not a typo: 27 million times!)

The highest time spent in a single function was 15.5% spent in `ucol_updateInternalState_52`.

**Comment by cperciva** [comment]
How many files do you have in `/usr/bin/` ?  Does valgrind give you a callgraph?

**Comment by gperciva** [comment]
```
$ ls /usr/bin/ | wc -l
3316
```

Looks like profiling Qt applications might need a different tool than the normal command-line valgrind; I'm looking into Qt creator valgrind stuff.

**Comment by gperciva** [comment]
Using Qt creator doesn't seem to give any more information.  It seems that `QCollatorPrivate::init()` owns the functions which get called a lot of times.  According to the docs, "The QCollator class compares strings according to a localized collation algorithm.".  Neither `QCollator` nor `QLocale` appear directly in the tarsnap-gui source code.

**Comment by cperciva** [comment]
Well, there is one very simple option... try running `killall -ABRT tarsnap-gui` while it's stuck.  See what the backtrace says. ;-)

**Comment by cperciva** [comment]
Hmm, I just tried on my FreeBSD system and I can't reproduce any slowness.  Maybe this was a bug in Qt which has since been fixed?

**Comment by gperciva** [comment]
hmm, I just tried on a fresh debian8 with no problems, even when running as the non-X11-owning user.  (i.e. `xhost local:@` ; `su another_user` ; `tarsnap-gui`)

I'm grabbing a fresh 14.04 iso to test inside virtualbox, but right now I'm leaning towards believing that my system has some weird locale thing going on.  I did a bunch of hackish stuff when I moved to Japan, and I was intending to wipe&reinstall the system when I left, so I didn't pay much attention to what hacks I was doing.

**Comment by gperciva** [comment]
on an almost-fresh Ubuntu 14.04 with 1524 files in `/usr/bin/`, adding `a`, or changing from `a` to `c` to `x` all result in approximately 1 second delays.  It's enough to be noticeable, but not the huge wait that I saw on my personal desktop.

**Comment by shinnok** [comment]
In your original case the locale mixup likely contributed to the long hangs.

That case aside, certainly there is a small processing spike and it's due to the Qt code that is listing the entire directory contents and parsing the metadata (icons, attributes) for populating the QTreeView. On multi core systems this should be unnoticeable, on single core it can be. I'll profile this and see if there is anything I can do for an improvement.

**Comment by shinnok** [comment]
I used an Ubuntu 14.04 LTS VM machine with a single core assigned to profile the File Browser performance as you did. The results conclude that 99.99% percent of the execution time is spent between the system libraries (LibICU - the I18N and locale library, libHarfbuzz - font processing, libc) and Qt SDK. Tarsnap GUI code only appears once and very low in the profiling result with the code in CustomFileSystemModel::data which, as stated previously, unless I have a really good reason to look at it again it's not worth even to attempt it. Its instruction count 498,010 in comparison with 2,125,188,751 for the top libICU cpu hog method is a whooping 4000x factor.

In conclusion, there isn't much I can do in Tarsnap GUI for improving this. Closing this issue until I have a reason to open it again.
